soluce_mentorat.py

```diff
@@ -26,20 +26,36 @@ def is_taken(self, x: int, y: int) -> bool:
         if status_position == Status.EMPTY or status_position == Status.OUT:
             return False
 
+        # Initialisation de l'historique des positions testées et de l'état de la forme
+        if positions is None:
+            positions = []
+            self.shape_is_free = False
+
         # 2ème étape: ❓ Est une position adjacente est libre ❓
         status_positions_adjacentes = [
-            self.get_status(x + 1, y),  # position adjacente ➡ droite
-            self.get_status(x - 1, y),  # position adjacente ⬅ gauche
-            self.get_status(x, y + 1),  # position adjacente ⬇ bas
-            self.get_status(x, y - 1),  # position adjacente ⬆ haute
+            (self.get_status(x + 1, y), (x + 1, y)),  # position adjacente ➡ droite
+            (self.get_status(x - 1, y), (x - 1, y)),  # position adjacente ⬅ gauche
+            (self.get_status(x, y + 1), (x, y + 1)),  # position adjacente ⬇ bas
+            (self.get_status(x, y - 1), (x, y - 1)),  # position adjacente ⬆ haute
         ]
         # 🔮 Est ce qu'une position adjacente est libre ❓
         for status_position_adjacente in status_positions_adjacentes:
             # ❓ Est ce que la position adjacente est libre ❓
-            if status_position_adjacente == Status.EMPTY:
-                # ✅ si oui la piece (x, y) n'est pas prise
-                return False
-        # ℹ Toute les positions adjacentes ne sont pas libres (i.e une pierre sur chaque position adjacente)
+            if status_position_adjacente[0] == Status.EMPTY:
+                # ✅ si oui la piece (x, y) est libre et donc la forme est libre
+                self.shape_is_free = True
+            # ❓ Sinon, est ce que les positions adjacentes de la position actuelle sont libres ❓
+            # On test une position adjacente seulement si:
+            # - Aucune position libre déjà trouvée (évite de continuer la recherche)
+            # - Pas déjà testé
+            # - La position actuelle est de la couleur de la position adjacente
+            elif self.shape_is_free == False and \
```

Comment:
`<bool_var> == False` ou `<bool_var> == True` sont des anti-patterns en Python.
Tu peux directement utiliser la variable booléenne (ou son [`not`](https://docs.python.org/3/reference/expressions.html#not)) à la place.